### PR TITLE
[PM-5741] Reworking usage of `window` object in BrowserApi.reloadExtension method

### PR DESCRIPTION
**Pull request by cagonzalezcs**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [x] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

The objective of these changes is to remove usage of the `window` object from the extension's background context, while applying a defensive approach to facilitating reloading the extension through a reload of the background script's location.

These changes shouldn't affect the behavior of the extension within manifest v2, but they do ensure that manifest v3 can call this method without running into any errors due to the service worker execution context.

## Code changes

- **apps/browser/src/background/main.background.ts:** Updating the reference to `BrowserApi.reloadExtension` to act upon the `self` global value rather than `window`.
- **apps/browser/src/platform/browser/browser-api.spec.ts:** Implementing Jest tests to verify changes made within `BrowserApi.reloadExtension`.
- **apps/browser/src/platform/browser/browser-api.ts:** Reworking `BrowserApi.reloadExtension` to check for the existence of the `window.location.reload` function when acting upon the passed global context.
- **apps/browser/test.setup.ts:** Adding a mock for `chrome.runtime.reload`

### CODE REVIEW

File: apps/browser/src/background/main.background.ts [hunk 1]
```diff
@@ -757,7 +757,7 @@ export default class MainBackground {
         this.platformUtilsService.isSafari() ||
         this.platformUtilsService.isFirefox() ||
         this.platformUtilsService.isOpera();
-      BrowserApi.reloadExtension(forceWindowReload ? window : null);
+      BrowserApi.reloadExtension(forceWindowReload ? self : null);
```
Comment (coroiu) [inline apps/browser/src/background/main.background.ts:760]: thought: TIL `self` is a global variable in service workers :)

File: apps/browser/src/platform/browser/browser-api.ts [hunk 2]
```diff
@@ -381,12 +381,20 @@ export class BrowserApi {
...
+   * to reload or by calling the extension's runtime to reload.
+   *
+   * @param globalContext - The global context to use for the reload.
+   */
+  static reloadExtension(globalContext: (Window & typeof globalThis) | null) {
+    // The passed globalContext might be a ServiceWorkerGlobalScope, as a result
+    // we need to check if the location object exists before calling reload on it.
+    if (typeof globalContext?.location?.reload === "function") {
+      return (globalContext as any).location.reload(true);
     }
```
Comment (coroiu) [inline apps/browser/src/platform/browser/browser-api.ts:395]: question: is `globalContext instanceof Window` not possible in this context?

**Review by coroiu — APPROVED** [review]
Looks good, just had one question

File: apps/browser/src/platform/browser/browser-api.ts (hunk 2 above)
Comment (cagonzalezcs) [reply apps/browser/src/platform/browser/browser-api.ts:395]: @coroiu 

That is possible to do, for sure. I've been avoiding usage of `instanceof` due to some performance issues that I've encountered within Firefox. I ran into a whole slew of issues with Firefox when reworking the autofill feature functionality (which heavily used `instanceof` to validate DOM object instance references). 

That said, in this cause it definitely would be less of a performance hit to call `globalContext instanceof Window`. Would you prefer that approach over checking explicitly for the presence of the `location.reload` method?

File: apps/browser/src/background/main.background.ts (hunk 1 above)
Comment (cagonzalezcs) [reply apps/browser/src/background/main.background.ts:760]: Ah yeah, it ends up referring to the `ServiceWorkerGlobalScope` - https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerGlobalScope

In the context of a DOM, `self` refers to the `Window`, so it's a valid global property for both instances.

File: apps/browser/src/platform/browser/browser-api.ts (hunk 2 above)
Comment (cagonzalezcs) [reply apps/browser/src/platform/browser/browser-api.ts:395]: I'm going to move forward with merging this in for now, if you have a strong enough concern with it I can update it in a future PR.

File: apps/browser/src/background/main.background.ts (hunk 1 above)
Comment (coroiu) [reply apps/browser/src/background/main.background.ts:760]: Oh I didn't know self worked in the regular browser page context, cool! :)

File: apps/browser/src/platform/browser/browser-api.ts (hunk 2 above)
Comment (coroiu) [reply apps/browser/src/platform/browser/browser-api.ts:395]: I didn't have any strong concerns, mostly curiosity. We're using this method to encapsulate the logic and abstract it away for the rest of the codebase, so it's easy to change if we ever reconsider :)

 Interesting to hear about the performance hit on Firefox, I'll keep that in mind going forward!

**Pull request merged**